### ISSUE TITLE
how to pass cookies to external graphql server

### ISSUE THREAD

**Issue by caedmon** [reactql/kit#77]
I've modified the auth example to use an external graphql server, and to only use a cookie for the JWT.  I ran into a problem with SSR in getting the JWT to the graphql server.  The only way I could manage to make it work was by modifying the createNetworkInterface() function in server.js to copy the cookie to the apollo network options.  Unfortunately middleware is called too late to do it there.

Is there a better way to do this that doesn't involve changes in the kit code?

**Comment by NiluK** [comment]
This is what I did:

// Server

```
js
config.addMiddleware(async (ctx, next) => {
    // Check the Authorization:` header or cookie for the JWT
    const authHeader = ctx.get('authorization') || ctx.cookies.get('reactQLJWT');
    if (authHeader) {
      // Strip out the `Bearer` prefix, so we're left with just the JWT and
      // store it on Koa's `ctx.state``.  At this point, it's unverified, but
      // we'll leave it to the relevant GraphQL query to do the validation
      // at the time of the data request, rather than taking up extra CPU cycles
      config.setApolloNetworkOptions({
        headers: {
          authorization: authHeader,
        },
        credentials: 'include',
      });
    }
    return next();
}
```

Browser 

```
js
config.addApolloMiddleware((req, next) => {
    const jwt = localStorage.getItem('reactQLJWT');
    if (req.options.headers) {
      req.options.headers = {
        ...req.options.headers,
        authorization: jwt || req.options.headers.authorization,
      };
    }
```

**Comment by leebenson** [comment]
@NiluK - be careful you're not running into edge cases with code above.

`config.setApolloNetworkOptions` changes the options for _all_ server rendered requests - so if you're attaching a request-specific auth header, that same header might be re-used on the _next_ request -- even if the next user hasn't authenticated!

The appropriate way to fix this would be to have a separate function -- say, `config.setApolloRequestOptions()`, along with a `config.addBeforeMiddleware()` function that can capture headers before `ctx.apollo` has been instantiated.

I'll consider how best to do this for the next release.

**Comment by NiluK** [comment]
@leebenson Good catch, that isn't something I was aware of, thanks. Is there a specific use case you're thinking of? I just double check it now with 4 different application caches (Chrome authenticated, Chrome incognito unauthenticated, Safari incognito unauthenticated, Firefox incognito unauthenticated + mobile versions) and they all seem to be handling correctly.

**Comment by leebenson** [comment]
I'm not sure if the above will _definitely_ happen, but I designed `config. setApolloNetworkOptions()` to be applied globally -- so it's worth checking that you're not accidentally passing options to the next request.

It's used by `getNetworkInterface()` in [kit/lib/apollo.js](https://github.com/reactql/kit/blob/master/kit/lib/apollo.js), which is called by `createNetworkInterface()` in [kit/entry/server.js](https://github.com/reactql/kit/blob/master/kit/entry/server.js) to attach to `ctx.apollo` on each new request.

So what **could** be happening is this:

1. User requests the site.
2. `ctx.apollo` gets the non-authenticated Apollo client first of all
3. User then authenticates in a future request, which runs the middleware and modifies the global Apollo network options
4. On the _next_ unauthenticated request, you have the previously set Apollo network options bound to `ctx.apollo` -- including the old cookie added as an authentication header.

You'd need to verify this by debugging (or just logging out to the console to be sure). I've not followed the code path myself, other than knowing that `config.setApolloNetworkOptions()` was never intended to be 'per-request'.

After I've added the aforementioned middleware functions, it should be safe to create a per-request Apollo client that definitely doesn't bleed out.

**Comment by NiluK** [comment]
@leebenson Excellent catch, thanks. That is indeed exactly what it does. I'll wait for your modified code before continuing further.

**Comment by caedmon** [comment]
@leebenson sounds like a good plan.  Thanks.

**Comment by NiluK** [comment]
@leebenson just wondering if there was any updates to this? Thanks.

**Comment by leebenson** [comment]
Not yet @NiluK. I put this on hold, pending https://github.com/reactql/kit/issues/80#issuecomment-332441041

I was initially considering making the switch to Apollo v2 before tackling any more userland API changes, but I'll probably wait a bit for the ecosystem to mature before swapping out the Apollo pieces. So the above should find its way into the next release.

**Comment by leebenson** [comment]
Features released in kit [2.9.1](https://github.com/reactql/kit/releases/tag/2.9.1) and [2.9.2](https://github.com/reactql/kit/releases/tag/2.9.2) should be enough to implement custom cookie handling/Apollo instantiation.

You can now do something like this:

```js
config.addBeforeMiddleware((ctx, next) => {
  ctx.apollo.networkOptions = {
    // some logic to set the per-request options
  };
  return next();
});
```

The Apollo client will instantiate automatically at `ctx.apollo.client` and will merge `ctx.apollo.options`(passed to `new ApolloClient()`) and `ctx.apollo.networkOptions` (in the call to `createNetworkInterface()`) when it does, so you can implement whatever header logic makes sense to your app.
